n3fit/src/n3fit/backends/keras_backend/MetaModel.py

```diff
@@ -167,12 +164,19 @@ def perform_fit(self, x=None, y=None, epochs=1, **kwargs):
         x_params = self._parse_input(x)
         if y is None:
             y = self.target_tensors
-        history = super().fit(x=x_params, y=y, epochs=epochs, **kwargs)
+
+        # This looks stupid, but it's actually faster, as it avoids some Tensorflow overhead
+        # every epoch. Each step is now a batch rather than an epoch
+        for k, v in x_params.items():
+            x_params[k] = tf.repeat(v, epochs, axis=0)
+        y = [tf.repeat(yi, epochs, axis=0) for yi in y]
```

Comment:
```suggestion
        # Instead of running # epochs, run #epochs batches in a single epoch
        # this avoids some Tensorflow overhead and it's actually faster
        for k, v in x_params.items():
            x_params[k] = tf.repeat(v, epochs, axis=0)
        y = [tf.repeat(yi, epochs, axis=0) for yi in y]
```

I'd say this is a clever trick, not a stupid one.

RE the memory usage, I wonder whether there's a way of tricking tensorflow to pass a tensor of length 1 in the batch dimension but make it believe there are actually many.
It depends on how the batch-taking is implemented in tensorflow but maybe we can do something like

```python
class FakeTensor(Tensor):
    def get_batch(self, i):
        if i < epochs:
             return self._true_tensor
         return end_signal
```

Reply:
I implemented something like this, it works, but unfortunately it seems to completely negate the benefits of the copying. No idea why.

Reply:
I think you might need to actually trick tensorflow for that to work.

However, I think the timings are quite ok as they are so we can leave this in the backburner for the time being. No need to overoptimize when there are many other bigger problems in the way.

Reply:
Well they seem ok here, but that's just because the rest is slow ;P It's about a factor 2 speedup once the other optimizations in issue #1803 are implemented.
But sure it can wait until the rest is done. I hoped to fix this quickly before I go on holiday, but of course it's always a bit more tricky than you expect. I think the simplest is to just make say 1000 copies and train for epochs/1000.

Reply:
> It's about a factor 2 speedup once the other optimizations in issue https://github.com/NNPDF/nnpdf/issues/1803 are implemented.

But is a factor 2 also in CPU? Because if it is only an improvement for GPU I'd say it's better to branch there. GPU and CPU are different enough devices that I think some branching is ok, like the `eigen`/`tensordot` thing and in CPU the memory growth can harm more than the 10% gain you quoted before.

Reply:
I don't know, it reduces the gap between steps from ~50 to ~3 ms, so it depends how much effect the other refactorings have on the CPU. If it's still not significant after that we can always default to the old behavior when the number of replicas is 1.